src/ast/mod.rs

```diff
@@ -2844,6 +2847,41 @@ impl fmt::Display for ListAggOnOverflow {
     }
 }
 
+/// An `ARRAY_AGG` invocation `ARRAY_AGG( [ DISTINCT ] <expr> [ORDER BY <expr>] [LIMIT <n>] )
+/// [ WITHIN GROUP (ORDER BY <within_group>  ]`
+/// ORDERY BY position is defined differently for BigQuery, Postgres and Snowflake
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct ArrayAgg {
+    pub distinct: bool,
+    pub expr: Box<Expr>,
+    pub order_by: Option<Box<OrderByExpr>>,
+    pub limit: Option<Box<Expr>>,
+    pub within_group: Option<Box<OrderByExpr>>,
```

Comment:
To be honest I'm not quite sure whether is good or not the usage of the order by inside the within group.

Considering the little difference, can't you just have a flag like `within_group` to identify if the order by is used inside a within group or not?

I couldn't find a situation where both are possible. Having both there make it possible to invalid (theoretically impossible) states to be built, which can be avoided in this case.

Reply:
I agree to your comment. In the beginning, I want to keep it like this for the purpose of easier to rewrite. But this can leads to invalid state. Will work on this.